File: consumer-app/src/stages.ts
// Five coarse conversion stages, shared visual language with the Android app.
export type Stage = 'import' | 'parse' | 'layout' | 'write' | 'done';
export const STAGES: { id: Stage; label: string }[] = [
  { id: 'import', label: 'Import' },
  { id: 'parse', label: 'Parse' },
  { id: 'layout', label: 'Layout' },
  { id: 'write', label: 'Write' },
  { id: 'done', label: 'Ready' },
];

export type StageHint = { stage: Stage; label: string; fraction: number };

// calibre emits log lines, not percentages. Map the lines we recognize to a
// stage + coarse fraction. Order is forward-only; the store never moves back.
const MARKERS: [string, StageHint][] = [
  ['imported browser_convert', { stage: 'import', label: 'Starting up', fraction: 0.06 }],
  ['Input running', { stage: 'parse', label: 'Reading the book', fraction: 0.18 }],
  ['Parsing all content', { stage: 'parse', label: 'Reading the book', fraction: 0.24 }],
  ['Merging user specified metadata', { stage: 'parse', label: 'Reading the book', fraction: 0.32 }],
  ['Detecting structure', { stage: 'layout', label: 'Laying out pages', fraction: 0.46 }],
  ['Detected chapters', { stage: 'layout', label: 'Laying out pages', fraction: 0.52 }],
  ['Flattening CSS', { stage: 'layout', label: 'Laying out pages', fraction: 0.58 }],
  ['Rasterizing', { stage: 'layout', label: 'Laying out pages', fraction: 0.62 }],
  ['inline TOC', { stage: 'layout', label: 'Laying out pages', fraction: 0.66 }],
  ['Creating', { stage: 'write', label: 'Building the Kindle file', fraction: 0.8 }],
  ['Serializing', { stage: 'write', label: 'Building the Kindle file', fraction: 0.86 }],
  ['Compressing', { stage: 'write', label: 'Building the Kindle file', fraction: 0.9 }],
  ['converted', { stage: 'done', label: 'Finishing', fraction: 0.98 }],
];

export function matchStage(line: string): StageHint | undefined {
  const lower = line.toLowerCase();
  for (const [needle, hint] of MARKERS) if (lower.includes(needle.toLowerCase())) return hint;
  return undefined;
}

export function stageIndex(stage: Stage): number {
  return Math.max(0, STAGES.findIndex((s) => s.id === stage));
}

File: android-app/app/src/main/kotlin/dev/exe/kindleconverter/convert/Stages.kt
package dev.exe.kindleconverter.convert

import dev.exe.kindleconverter.data.Stage

/** A coarse, honest progress point derived from a calibre log line. */
data class StageHint(val stage: Stage, val label: String, val fraction: Float)

// calibre emits log lines, not percentages. We map the lines we know to a 5-stage rail
// with a coarse fraction. Order matters: later matches represent later progress; the
// manager only ever moves forward, so noisy ordering can't make the bar jump backward.
private val MARKERS = listOf(
    "imported browser_convert" to StageHint(Stage.IMPORT, "Starting up", 0.06f),
    "Input running" to StageHint(Stage.PARSE, "Reading the book", 0.18f),
    "Parsing all content" to StageHint(Stage.PARSE, "Reading the book", 0.24f),
    "Merging user specified metadata" to StageHint(Stage.PARSE, "Reading the book", 0.32f),
    "Detecting structure" to StageHint(Stage.LAYOUT, "Laying out pages", 0.46f),
    "Detected chapters" to StageHint(Stage.LAYOUT, "Laying out pages", 0.52f),
    "Flattening CSS" to StageHint(Stage.LAYOUT, "Laying out pages", 0.58f),
    "Rasterizing" to StageHint(Stage.LAYOUT, "Laying out pages", 0.62f),
    "inline TOC" to StageHint(Stage.LAYOUT, "Laying out pages", 0.66f),
    "Creating" to StageHint(Stage.WRITE, "Building the Kindle file", 0.80f),
    "Serializing" to StageHint(Stage.WRITE, "Building the Kindle file", 0.86f),
    "Compressing" to StageHint(Stage.WRITE, "Building the Kindle file", 0.90f),
    "converted" to StageHint(Stage.DONE, "Finishing", 0.98f),
    "Output saved" to StageHint(Stage.DONE, "Finishing", 0.99f),
)

fun matchStage(line: String): StageHint? =
    MARKERS.firstOrNull { line.contains(it.first, ignoreCase = true) }?.second

/** A representative fill for a stage, used when no live fraction is available (e.g. the list). */
fun stageBaseFraction(stage: Stage) = when (stage) {
    Stage.IMPORT -> 0.08f
    Stage.PARSE -> 0.30f
    Stage.LAYOUT -> 0.58f
    Stage.WRITE -> 0.85f
    Stage.DONE -> 1f
}
